metricshub-agent/src/main/java/org/metricshub/web/service/ConfigurationFilesService.java

```diff
@@ -0,0 +1,334 @@
+	 */
+	private static final Set<String> YAML_EXTENSIONS = Set.of(".yml", ".yaml");
+
+	/**
+	 * Maximum directory traversal depth when searching for configuration files.
+	 */
+	private static final int MAX_DEPTH = 1;
+
+	private final AgentContextHolder agentContextHolder;
+	private final Yaml yaml = new Yaml();
+
+	@Autowired
+	public ConfigurationFilesService(final AgentContextHolder agentContextHolder) {
+		this.agentContextHolder = agentContextHolder;
+	}
+
+	/**
+	 * Retrieves a list of all configuration files with their metadata.
+	 *
+	 * @return A list of {@link ConfigurationFile} representing all configuration
+	 *         files.
+	 */
+	public List<ConfigurationFile> getAllConfigurationFiles() {
+		final var agentContext = agentContextHolder.getAgentContext();
+		if (agentContext != null) {
+			final var configurationDirectory = agentContext.getConfigDirectory();
+			// List all the files
+			try (
+				Stream<Path> files = Files.find(configurationDirectory, MAX_DEPTH, ConfigurationFilesService::isRegularYamlFile)
+			) {
+				return files
+					.map(this::buildConfigurationFile)
+					.filter(Objects::nonNull)
+					.sorted(Comparator.comparing(ConfigurationFile::getName, String.CASE_INSENSITIVE_ORDER))
+					.toList();
+			} catch (Exception e) {
+				log.error("Failed to list configuration directory: '{}'. Error: {}", configurationDirectory, e.getMessage());
+				log.debug("Failed to list configuration directory: '{}'. Exception:", configurationDirectory, e);
+			}
+		}
+		return List.of();
+	}
+
+	/**
+	 * Returns the content of a configuration file as UTF-8 text.
+	 *
+	 * @param fileName the configuration file name (e.g. metricshub.yaml)
+	 * @return file content
+	 */
+	public String getFileContent(final String fileName) {
+		final Path dir = requireConfigDir();
+		final Path file = resolveSafeYaml(dir, fileName);
+		log.info("Reading config file: {}", file.toAbsolutePath());
+
+		if (!Files.exists(file)) {
+			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Configuration file not found.");
+		}
+		try {
+			return Files.readString(file, StandardCharsets.UTF_8);
+		} catch (IOException e) {
+			log.error("Failed to read configuration file: '{}'. Error: {}", file, e.getMessage());
+			log.debug("Failed to read configuration file: '{}'. Exception:", file, e);
+			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to read configuration file.", e);
+		}
+	}
+
+	/**
+	 * Saves or updates the content of a configuration file (atomic write).
+	 *
+	 * @param fileName the configuration file name
+	 * @param content  the content to write
+	 */
+	public void saveOrUpdateFile(final String fileName, final String content) {
+		final Path dir = requireConfigDir();
+		final Path target = resolveSafeYaml(dir, fileName);
+
+		try {
+			Files.createDirectories(dir);
+			// Write to a temp file, then move atomically
+			final Path tmp = Files.createTempFile(dir, fileName + ".", ".tmp");
+			Files.writeString(

```

Comment:
Before saving the file, we should probably validate the file before saving it, and throw an exception if the file is invalid. If we save an invalid file, it will prevent the product from working.

Maybe we could add an option to skip the validation on save (`PUT /api/config-files/my-config.yaml?skipValidation=true`).